File: kb30-wear/shared/src/main/kotlin/nl/kb30/protocol/SwingLedger.kt
package nl.kb30.protocol

/**
 * Conflict-free swing counter.
 *
 * Acceptance criterion: "swing counter syncs without double counting", and
 * "temporary connection loss causes no lost session". A naive `count += 1` on
 * each side breaks the moment the watch and phone reconnect and re-exchange
 * totals. So instead every tap is a distinct, immutable **event** identified by
 * `"$origin:$seq"`, and the ledger is a small CRDT (an OR/2P-set):
 *
 *  - [added]   : every swing event ever recorded (grow-only).
 *  - [removed] : tombstones for events undone by the user (grow-only).
 *  - count     = |added \ removed|.
 *
 * Merging two ledgers is just the union of both sets on both fields, which is
 * commutative, associative and idempotent — so any order or number of reconnect
 * exchanges converges to the same count, with no double counting and no lost
 * undo. Sequence numbers are monotonic per origin, so an undone event is never
 * resurrected by a later merge.
 */
class SwingLedger private constructor(
    private val added: MutableSet<String>,
    private val removed: MutableSet<String>,
) {
    constructor() : this(LinkedHashSet(), LinkedHashSet())

    /** Current, de-duplicated swing count. */
    val count: Int
        get() = added.count { it !in removed }

    /** Next unused sequence number for [origin]. */
    private fun nextSeq(origin: String): Long {
        var max = 0L
        val prefix = "$origin:"
        for (id in added) {
            if (id.startsWith(prefix)) {
                val seq = id.substring(prefix.length).toLongOrNull() ?: continue
                if (seq > max) max = seq
            }
        }
        return max + 1
    }

    /** Record one swing tap from [origin]. Returns the event id created. */
    fun add(origin: String): String {
        val id = "$origin:${nextSeq(origin)}"
        added.add(id)
        return id
    }

    /**
     * Undo the most recent still-active swing from [origin] (matches the watch's
     * per-device undo button). Returns the tombstoned event id, or null if there
     * was nothing to undo for that origin.
     */
    fun undoLast(origin: String): String? {
        val prefix = "$origin:"
        val candidate = added
            .asSequence()
            .filter { it.startsWith(prefix) && it !in removed }
            .maxByOrNull { it.substring(prefix.length).toLongOrNull() ?: -1 }
            ?: return null
        removed.add(candidate)
        return candidate
    }

    /** Fold another ledger into this one. Commutative / idempotent. */
    fun merge(other: SwingLedger): SwingLedger {
        added.addAll(other.added)
        removed.addAll(other.removed)
        return this
    }

    fun copy(): SwingLedger =
        SwingLedger(LinkedHashSet(added), LinkedHashSet(removed))

    fun toMap(): Map<String, Any?> = linkedMapOf(
        "added" to added.toList(),
        "removed" to removed.toList(),
    )

    fun toJson(): String = Json.encode(toMap())

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromMap(m: Map<String, Any?>): SwingLedger {
            val added = (m["added"] as? List<Any?>)?.map { it.asStr() } ?: emptyList()
            val removed = (m["removed"] as? List<Any?>)?.map { it.asStr() } ?: emptyList()
            return SwingLedger(LinkedHashSet(added), LinkedHashSet(removed))
        }

        fun fromJson(json: String): SwingLedger =
            try { fromMap(Json.decodeObject(json)) } catch (_: Exception) { SwingLedger() }
    }
}

File: kb30-pwa/js/swingLedger.js
// swingLedger.js — conflictvrije swingteller (CRDT), tegenhanger van de Kotlin
// SwingLedger. Elke tik is een uniek event "origin:seq". De teller is een
// 2P-set: added (alle tikken) en removed (undo's). count = |added \ removed|.
// Merge = union op beide velden → geen dubbeltellingen, undo's overleven merges.

export class SwingLedger {
  constructor(added = [], removed = []) {
    this.added = new Set(added);
    this.removed = new Set(removed);
  }

  get count() {
    let n = 0;
    for (const id of this.added) if (!this.removed.has(id)) n++;
    return n;
  }

  _nextSeq(origin) {
    const prefix = origin + ':';
    let max = 0;
    for (const id of this.added) {
      if (id.startsWith(prefix)) {
        const seq = parseInt(id.slice(prefix.length), 10);
        if (Number.isFinite(seq) && seq > max) max = seq;
      }
    }
    return max + 1;
  }

  add(origin) {
    const id = origin + ':' + this._nextSeq(origin);
    this.added.add(id);
    return id;
  }

  undoLast(origin) {
    const prefix = origin + ':';
    let best = null;
    let bestSeq = -1;
    for (const id of this.added) {
      if (id.startsWith(prefix) && !this.removed.has(id)) {
        const seq = parseInt(id.slice(prefix.length), 10);
        if (seq > bestSeq) { bestSeq = seq; best = id; }
      }
    }
    if (best == null) return null;
    this.removed.add(best);
    return best;
  }

  merge(other) {
    for (const id of other.added) this.added.add(id);
    for (const id of other.removed) this.removed.add(id);
    return this;
  }

  copy() {
    return new SwingLedger([...this.added], [...this.removed]);
  }

  toJSON() {
    return { added: [...this.added], removed: [...this.removed] };
  }

  static fromJSON(obj) {
    try {
      const o = typeof obj === 'string' ? JSON.parse(obj) : obj;
      return new SwingLedger(o.added || [], o.removed || []);
    } catch {
      return new SwingLedger();
    }
  }
}
